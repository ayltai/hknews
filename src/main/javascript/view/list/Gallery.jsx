import PropTypes from 'prop-types';
import React from 'react';
import { SingleFieldList, } from 'react-admin';
import Lightbox from 'react-image-lightbox';
import 'react-image-lightbox/style.css';

import { StringUtils, } from '../../util/StringUtils';

const getEntries = props => props.ids && props.data ? props.ids.map(id => props.data[id]) : [];

const getImages = props => Object.entries(getEntries(props)).map(record => ({
    label  : StringUtils.decode(record[1][props.caption]),
    source : record[1][props.source],
}));

export const Gallery = props => {
    const [ index, setIndex, ]      = React.useState(0);
    const [ open,  setOpen,  ]      = React.useState(false);
    const { source, ...otherProps } = props;
    const entries                   = getEntries(props);
    const images                    = getImages(props);

    const children = React.Children.map(props.children, child => React.cloneElement(child, {
        onClick : event => {
            setIndex(Math.max(0, entries ? entries.findIndex(entry => entry[source] === event.target.src) : 0));
            setOpen(true);
        },
    }));

    return (
        <React.Fragment>
            <SingleFieldList
                linkType={false}
                {...otherProps}>
                {children[0]}
            </SingleFieldList>
            {open && (
                <Lightbox
                    mainSrc={images[index].source}
                    nextSrc={index + 1 < images.length ? images[(index + 1) % images.length].source : undefined}
                    prevSrc={index - 1 >= 0 ? images[(index + images.length - 1) % images.length].source : undefined}
                    imageCaption={images[index].label}
                    onCloseRequest={() => setOpen(false)}
                    onMoveNextRequest={() => setIndex((index + 1) % images.length)}
                    onMovePrevRequest={() => setIndex((index + images.length - 1) % images.length)} />
            )}
        </React.Fragment>
    );
};

Gallery.propTypes = {
    ids      : PropTypes.arrayOf(PropTypes.number),
    data     : PropTypes.object,
    caption  : PropTypes.string,
    source   : PropTypes.string,
    children : PropTypes.object,
};
