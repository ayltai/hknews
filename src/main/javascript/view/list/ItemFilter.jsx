import React from 'react';
import { Filter, ReferenceInput, SelectInput, } from 'react-admin';

export const ItemFilter = props => (
    <Filter {...props}>
        <ReferenceInput
            label='labels.publisher'
            source='source'
            reference='sources'
            allowEmpty
            alwaysOn>
            <SelectInput
                optionText='name'
                variant='standard' />
        </ReferenceInput>
    </Filter>
);
