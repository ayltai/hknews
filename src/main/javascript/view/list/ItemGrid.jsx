import { Card, CardActions, CardActionArea, CardContent, CardHeader, CardMedia, Grid, makeStyles, Tooltip, } from '@material-ui/core';
import PropTypes from 'prop-types';
import React from 'react';
import { RichTextField, ShowButton, TextField, } from 'react-admin';
import TimeAgo from 'react-timeago';

import { StringUtils, } from '../../util/StringUtils';
import { Constants, } from '../../Constants';

export const ItemGrid = ({ basePath, ids, data, }) => {
    const classes = makeStyles({
        container   : {
            padding  : 16,
            flexGrow : 1,
        },
        title       : {
            paddingBottom : '0.75em',
            display       : 'block',
            overflow      : 'hidden',
            fontSize      : '120%',
            fontWeight    : 'bold',
            textOverflow  : 'ellipsis',
            whiteSpace    : 'nowrap',
        },
        description : {
            minHeight       : 200,
            display         : '-webkit-box',
            overflow        : 'hidden',
            textOverflow    : 'ellipsis',
            WebkitBoxOrient : 'vertical',
            WebkitLineClamp : 8,
        },
        media       : {
            height     : 0,
            paddingTop : '56.25%',
        },
        actions     : {
            display   : 'block',
            textAlign : 'right',
        },
    })();

    return (
        <div className={classes.container}>
            <Grid
                container
                spacing={2}>
                {ids.map(id => (
                    <Grid
                        item
                        key={id}
                        xs={12}
                        sm={6}
                        md={6}
                        lg={4}
                        xl={3}>
                        <Card>
                            <CardHeader
                                title={
                                    <TextField
                                        record={data[id]}
                                        source='sourceName' />
                                }
                                subheader={<TimeAgo date={data[id].publishDate} />}
                                avatar={
                                    <img
                                        src={`${Constants.API_ENDPOINT}/images/${Constants.SOURCE_IMAGES[data[id].sourceName]}.png`}
                                        width={Constants.AVATAR_SIZE}
                                        height={Constants.AVATAR_SIZE}
                                        title={data[id].sourceName}
                                        alt={data[id].sourceName} />
                                } />
                            <CardActionArea href={`#${basePath}/${id}/show`}>
                                {data[id].images && data[id].images.length > 0 && (
                                    <Tooltip title={data[id].images[0].description ? StringUtils.decode(data[id].images[0].description) : ''}>
                                        <CardMedia
                                            className={classes.media}
                                            image={data[id].images[0].imageUrl} />
                                    </Tooltip>
                                )}
                                <CardContent>
                                    <RichTextField
                                        className={classes.title}
                                        record={data[id]}
                                        source='title' />
                                    <RichTextField
                                        className={classes.description}
                                        record={data[id]}
                                        source='description' />
                                </CardContent>
                            </CardActionArea>
                            <CardActions className={classes.actions}>
                                <ShowButton
                                    basePath={basePath}
                                    label='labels.read_more'
                                    record={data[id]} />
                            </CardActions>
                        </Card>
                    </Grid>
                ))}
            </Grid>
        </div>
    );
};

ItemGrid.propTypes = {
    basePath : PropTypes.string,
    ids      : PropTypes.arrayOf(PropTypes.string),
    data     : PropTypes.object,
};
