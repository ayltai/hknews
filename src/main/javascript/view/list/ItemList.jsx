import React from 'react';
import { List, Pagination, } from 'react-admin';

import { Constants, } from '../../Constants';
import { ScrollToTopOnMount, } from '../ScrollToTopOnMount';
import { ItemFilter, } from './ItemFilter';
import { ItemGrid, } from './ItemGrid';

export const ItemList = props => (
    <React.Fragment>
        <ScrollToTopOnMount />
        <List
            bulkActionButtons={false}
            exporter={false}
            filters={<ItemFilter />}
            perPage={Constants.NEWS_PER_PAGE}
            pagination={<Pagination rowsPerPageOptions={Constants.NEWS_PER_PAGE_OPTIONS} />}
            {...props}>
            <ItemGrid />
        </List>
    </React.Fragment>
);
