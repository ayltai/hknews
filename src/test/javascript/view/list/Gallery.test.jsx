import { mount, } from 'enzyme';
import React from 'react';
import { TestContext, } from 'react-admin';
import renderer from 'react-test-renderer';

import { Gallery, } from '../../../../main/javascript/view/list/Gallery';

describe('<Gallery />', () => {
    const component = (
        <TestContext>
            <Gallery
                ids={[ 'id1', 'id2', ]}
                data={{
                    id1 : {
                        label  : '1',
                        source : 'image1',
                    },
                    id2 : {
                        label  : '2',
                        source : 'image2',
                    },
                }}
                label='label'
                source='source'>
                <img src='image1' />
            </Gallery>
        </TestContext>
    );

    it('renders without errors', () => {
        expect(renderer.create(component).toJSON()).toMatchSnapshot();
    });

    it('mounts without errors', () => {
        mount(component);
    });
});
