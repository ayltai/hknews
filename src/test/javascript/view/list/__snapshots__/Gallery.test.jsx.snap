// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Gallery /> renders without errors 1`] = `
<div
  className="RaSingleFieldList-root-1"
  label="label"
>
  <img
    onClick={[Function]}
    record={
      Object {
        "label": "1",
        "source": "image1",
      }
    }
    src="image1"
  />
  <img
    onClick={[Function]}
    record={
      Object {
        "label": "2",
        "source": "image2",
      }
    }
    src="image1"
  />
</div>
`;
